File: src/core/models/schedule.ts
/**
 * Schedule — a persisted cron-based agent trigger.
 *
 * When a schedule fires, it creates a BackgroundTask which runs the
 * configured ACP agent with the specified prompt.
 */

export interface Schedule {
  id: string;
  /** Human-readable name */
  name: string;
  /** Cron expression (standard 5-field: min hour dom mon dow) */
  cronExpr: string;
  /** Prompt to dispatch to the agent when the schedule fires */
  taskPrompt: string;
  /** ACP agent / provider ID (e.g. "claude-code", "opencode") */
  agentId: string;
  /** Workspace the schedule belongs to */
  workspaceId: string;
  /** Whether this schedule is active */
  enabled: boolean;
  /** Timestamp of the last successful trigger */
  lastRunAt?: Date;
  /** Computed next trigger time */
  nextRunAt?: Date;
  /** Background task ID from the most recent trigger */
  lastTaskId?: string;
  /** Optional prompt template override (supports {timestamp}, {cronExpr}) */
  promptTemplate?: string;
  createdAt: Date;
  updatedAt: Date;
}

/**
 * Input required to create a new schedule.
 */
export interface CreateScheduleInput {
  id?: string;
  name: string;
  cronExpr: string;
  taskPrompt: string;
  agentId: string;
  workspaceId: string;
  enabled?: boolean;
  promptTemplate?: string;
}

/**
 * Build the prompt text from a schedule, substituting template variables.
 */
export function resolveSchedulePrompt(schedule: Schedule): string {
  const template = schedule.promptTemplate?.trim() || schedule.taskPrompt;
  return template
    .replace(/\{timestamp\}/g, new Date().toISOString())
    .replace(/\{cronExpr\}/g, schedule.cronExpr)
    .replace(/\{scheduleName\}/g, schedule.name);
}

File: crates/routa-core/src/models/schedule.rs
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};

/// A cron-based scheduled agent trigger.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Schedule {
    pub id: String,
    pub name: String,
    pub cron_expr: String,
    pub task_prompt: String,
    pub agent_id: String,
    pub workspace_id: String,
    pub enabled: bool,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub last_run_at: Option<DateTime<Utc>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub next_run_at: Option<DateTime<Utc>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub last_task_id: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub prompt_template: Option<String>,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

/// Input for creating a new schedule.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CreateScheduleInput {
    pub name: String,
    pub cron_expr: String,
    pub task_prompt: String,
    pub agent_id: String,
    pub workspace_id: String,
    #[serde(default = "default_true")]
    pub enabled: bool,
    pub next_run_at: Option<DateTime<Utc>>,
    pub prompt_template: Option<String>,
}

fn default_true() -> bool {
    true
}

/// Partial update input for PATCH.
#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct UpdateScheduleInput {
    pub name: Option<String>,
    pub cron_expr: Option<String>,
    pub task_prompt: Option<String>,
    pub agent_id: Option<String>,
    pub enabled: Option<bool>,
    pub next_run_at: Option<DateTime<Utc>>,
    pub last_run_at: Option<DateTime<Utc>>,
    pub last_task_id: Option<String>,
    pub prompt_template: Option<String>,
}
